### DISCUSSION TITLE
Vercel Next JS Cache MISS and HIT questions

### DISCUSSION THREAD

**Discussion by chrissshh** [vercel/next.js#67732] [Help]
I have a dynamic page in Next JS 14 located at "/product/[slug]" and I'm having trouble understanding how to reduce the serverless function invocations in vercel since most of these pages are generated and cached for 3 months. I always get Cache MISS and the function runs even if the data is cached for a specific slug. 

My Question: Does a dynamic route always require a function invocation even if the data is cached?

Here is sample code of the page
```
export default async function ProductDetails({
  params,
}: {
  params: { slug: string };
}) {
  const res = await fetch(
   "https://www.another-domain.com/api/product?p=${params.slug}",
    {
      method: "GET",
      next: { revalidate: 60000000 },
    },
  ).then((response) => response.json());

  return (
    <div className="">
      <p>{JSON.stringify(res,null,2)}</p>
    </div>
  );
}
```

**Comment by HakanSungur** [comment]
I've same problem.

**Comment by DiogoS** [comment]
Did you managed to find a solution?

**Comment by momanamjad** [comment]
Great question! The Cache MISS behavior you're seeing is actually how ISR (Incremental 
Static Regeneration) works in Next.js 14. Here's what's happening:

**Understanding the Caching Behavior:**

1. **Dynamic routes always trigger a function invocation on first request** — this is expected 
   behavior. The `revalidate` option on your fetch() controls DATA caching, not page generation.

2. **Your revalidate time might be too aggressive** — you're setting:
```javascript
   next: { revalidate: 60000000 } // ~2 years in seconds
```
   This should work, but the issue might be that each unique `[slug]` is treated separately.

**The Real Issue - On-Demand ISR:**
Each dynamic route slug is generated on-demand the first time it's requested. After that, 
it should be cached for your revalidate window. However, on Vercel, there are a few reasons 
you might see Cache MISS:

- **Different deployment or cache cleared** — Vercel clears cache on new deployments
- **Regional caching** — Different regions might have separate caches
- **Stale-While-Revalidate** — The page might be serving stale content while revalidating 
  in the background

**Solutions:**

1. **Use revalidateTag() for manual control:**
```javascript
   export default async function ProductDetails({ params }: Props) {
     const res = await fetch(
       `https://www.another-domain.com/api/product?p=${params.slug}`,
       { 
         method: 'GET',
         next: { tags: ['product', params.slug] } // Use tags instead
       }
     ).then(r => r.json());
     
     return <div><p>{JSON.stringify(res, null, 2)}</p></div>;
   }
```

2. **On Vercel, check your ISR timeout** — increase it if needed via `vercel.json`:
```json
   {
     "functions": {
       "app/product/[slug]/page.js": {
         "maxDuration": 60
       }
     }
   }
```

3. **Consider static generation for known slugs:**
```javascript
   export async function generateStaticParams() {
     // Pre-generate popular products at build time
     return [{ slug: 'product-1' }, { slug: 'product-2' }];
   }
```

The Cache MISS you're seeing is likely because each new slug hasn't been generated yet. 
After the first request for a slug, it should be cached for your revalidate period. Check 
your Vercel Analytics to confirm if requests are hitting the cache on subsequent requests!
